**To scrape the given Wikipedia page and create a Dataframe**

In [2]:
!pip install BeautifulSoup4

     |████████████████████████████████| 112kB 31.6MB/s eta 0:00:01


In [3]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

In [5]:
!pip install lxml

     |████████████████████████████████| 5.8MB 7.7MB/s eta 0:00:01


In [7]:
!pip install html5lib

In [9]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(url).text
Canada_data = BeautifulSoup(source, 'html.parser')

In [10]:
column_names = ['Postalcode','Borough','Neighborhood']
toronto = pd.DataFrame(columns = column_names)

In [11]:
content = Canada_data.find('div', class_='mw-parser-output')
table = content.table.tbody
postcode = 0
borough = 0
neighborhood = 0

for tr in table.find_all('tr'):
    i = 0
    for td in tr.find_all('td'):
        if i == 0:
            postcode = td.text
            i = i + 1
        elif i == 1:
            borough = td.text
            i = i + 1
        elif i == 2: 
            neighborhood = td.text.strip('\n').replace(']','')
    toronto = toronto.append({'Postalcode': postcode,'Borough': borough,'Neighborhood': neighborhood},ignore_index=True)

# clean dataframe 
toronto = toronto[toronto.Borough!='Not assigned']
toronto = toronto[toronto.Borough!= 0]
toronto.reset_index(drop = True, inplace = True)
i = 0
for i in range(0,toronto.shape[0]):
    if toronto.iloc[i][2] == 'Not assigned':
        toronto.iloc[i][2] = toronto.iloc[i][1]
        i = i+1
                                 
df = toronto.groupby(['Postalcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df.head()

Postalcode      Borough                            Neighborhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [13]:
df = df.dropna()
empty = 'Not assigned'
df = df[(df.Postalcode != empty ) & (df.Borough != empty) & (df.Neighborhood != empty)]

In [14]:
df.head()

Postalcode      Borough                            Neighborhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [15]:
def neighborhood_list(grouped):    
    return ', '.join(sorted(grouped['Neighborhood'].tolist()))
                    
grp = df.groupby(['Postalcode', 'Borough'])
df2 = grp.apply(neighborhood_list).reset_index(name='Neighborhood')

In [16]:
print(df2.shape)
df2.head(10)

(103, 3)


Postalcode      Borough                                     Neighborhood
0        M1B  Scarborough                                   Rouge, Malvern
1        M1C  Scarborough           Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough                Guildwood, Morningside, West Hill
3        M1G  Scarborough                                           Woburn
4        M1H  Scarborough                                        Cedarbrae
5        M1J  Scarborough                              Scarborough Village
6        M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park
7        M1L  Scarborough                  Clairlea, Golden Mile, Oakridge
8        M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West
9        M1N  Scarborough                      Birch Cliff, Cliffside West

In [17]:
print('The DataFrame shape is', df2.shape)

The DataFrame shape is (103, 3)



The dataframe has 103 Postal codes but it has 212 rows, because each Postal code can present more than one neighborhood (210 in total). Therefore, the dataframe should be group by the Postal code, ending with a dataframe with 103 rows.

**Task 2**

In [18]:
!wget -q -O 'Toronto_long_lat_data.csv'  http://cocl.us/Geospatial_data
df_lon_lat = pd.read_csv('Toronto_long_lat_data.csv')
df_lon_lat.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [19]:
df_lon_lat.columns=['Postalcode','Latitude','Longitude']

In [20]:
df_lon_lat.head()

Postalcode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [21]:
Toronto_df = pd.merge(df2,
                 df_lon_lat[['Postalcode','Latitude', 'Longitude']],
                 on='Postalcode')
Toronto_df

Postalcode      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
..         ...          ...   
98         M9N         York   
99         M9P    Etobicoke   
100        M9R    Etobicoke   
101        M9V    Etobicoke   
102        M9W    Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, Martin Grove Gardens, Richv...  43.688905 -79.554724  
101  Albion Gardens, Beaumond Heights, Humbergate, ...  43.739416 -79.588437  
102                                          Northwest  43.706748 -79.594054  

[103 rows x 5 columns]

**Task3**

In [23]:
!pip install geopy

     |████████████████████████████████| 112kB 13.5MB/s eta 0:00:01


In [24]:
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [25]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="Toronto")
location = geolocator.geocode(address)
latitude_toronto = location.latitude
longitude_toronto = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude_toronto, longitude_toronto))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [28]:
map_toronto = folium.Map(location=[latitude_toronto, longitude_toronto], zoom_start=10)

# add markers to map
for lat, lng, borough, Neighborhood in zip(Toronto_df['Latitude'], Toronto_df['Longitude'], Toronto_df['Borough'], Toronto_df['Neighborhood']):
    label = '{}, {}'.format(Neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [29]:
# defining radius and limit of venues to get
radius=500
LIMIT=100

In [30]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [34]:
CLIENT_ID = 'OCEPD1EW4CBKDMFQEHKKZCZSZSKJSRJZFUUXUB321HXFSVPN' # your Foursquare ID
CLIENT_SECRET = '5CC4MQ22RTNZRVD4BSWFA5OC4MCULKTGDD4QXF5KOZH41CIE' # your Foursquare Secret
VERSION = '20180604'

In [35]:
toronto_venues = getNearbyVenues(names=Toronto_df['Neighborhood'],
                                   latitudes=Toronto_df['Latitude'],
                                   longitudes=Toronto_df['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

In [36]:
toronto_venues.head(10)

Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1  Highland Creek, Rouge Hill, Port Union              43.784535   
2  Highland Creek, Rouge Hill, Port Union              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   
5       Guildwood, Morningside, West Hill              43.763573   
6       Guildwood, Morningside, West Hill              43.763573   
7       Guildwood, Morningside, West Hill              43.763573   
8       Guildwood, Morningside, West Hill              43.763573   
9       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                         Venue  Venue Latitude  \
0              -79.194353                       Wendy's       43.807448   
1              -79.160497         Royal Canadian Legion       43.782533   
2              -79.160497     Affordable Toronto Movers       43.787919   
3              -79.188711             G & G Electronics       43.765309   
4              -79.188711                    Marina Spa       43.766000   
5              -79.188711              Big Bite Burrito       43.766299   
6              -79.188711         Enterprise Rent-A-Car       43.764076   
7              -79.188711         Woburn Medical Centre       43.766631   
8              -79.188711  Lawrence Ave E & Kingston Rd       43.767704   
9              -79.188711                      Eggsmart       43.767800   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.162977         Moving Target  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa  
5       -79.190720    Mexican Restaurant  
6       -79.193406   Rental Car Location  
7       -79.192286        Medical Center  
8       -79.189490          Intersection  
9       -79.190466        Breakfast Spot

In [37]:
toronto_venues.shape

(2239, 7)

In [38]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                     11   
Alderwood, Long Branch                                                  9   
...                                                                   ...   
Willowdale West                                                         7   
Woburn                                                                  3   
Woodbine Gardens, Parkview Hill                                        12   
Woodbine Heights                                                        7   
York Mills West                                                         4   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Agincourt                                                                4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                       2   
Albion Gardens, Beaumond Heights, Humbergate, J...                      11   
Alderwood, Long Branch                                                   9   
...                                                                    ...   
Willowdale West                                                          7   
Woburn                                                                   3   
Woodbine Gardens, Parkview Hill                                         12   
Woodbine Heights                                                         7   
York Mills West                                                          4   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                              100             100   
Agincourt                                               4               4   
Agincourt North, L'Amoreaux East, Milliken, Ste...      2               2   
Albion Gardens, Beaumond Heights, Humbergate, J...     11              11   
Alderwood, Long Branch                                  9               9   
...                                                   ...             ...   
Willowdale West                                         7               7   
Woburn                                                  3               3   
Woodbine Gardens, Parkview Hill                        12              12   
Woodbine Heights                                        7               7   
York Mills West                                         4               4   

                                                    Venue Longitude  \
Neighborhood                                                          
Adelaide, King, Richmond                                        100   
Agincourt                                                         4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                2   
Albion Gardens, Beaumond Heights, Humbergate, J...               11   
Alderwood, Long Branch                                            9   
...                                                             ...   
Willowdale West                                                   7   
Woburn                                                            3   
Woodbine Gardens, Parkview Hill                                  12   
Woodbine Heights                                                  7   
York Mills West                                                   4   

                                                    Venue Category  
Neighborhood                            

**Analysing Each Neighborhood**

In [39]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot.head()

Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                  0                  0        0                   0   
1                  0                  0        0                   0   
2                  0                  0        0                   0   
3                  0                  0        0                   0   
4                  0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Vegetarian / Vegan Restaurant  \
0                    0             0  ...                              0   
1                    0             0  ...                              0   
2                    0             0  ...                              0   
3                    0             0  ...                              0   
4                    0             0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 268 columns]

In [40]:
toronto_onehot.shape

(2239, 268)

In [42]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                            Adelaide, King, Richmond                0.0   
1                                           Agincourt                0.0   
2   Agincourt North, L'Amoreaux East, Milliken, St...                0.0   
3   Albion Gardens, Beaumond Heights, Humbergate, ...                0.0   
4                              Alderwood, Long Branch                0.0   
..                                                ...                ...   
95                                    Willowdale West                0.0   
96                                             Woburn                0.0   
97                    Woodbine Gardens, Parkview Hill                0.0   
98                                   Woodbine Heights                0.0   
99                                    York Mills West                0.0   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                 0.0      0.0                 0.0           0.0   
1                 0.0      0.0                 0.0           0.0   
2                 0.0      0.0                 0.0           0.0   
3                 0.0      0.0                 0.0           0.0   
4                 0.0      0.0                 0.0           0.0   
..                ...      ...                 ...           ...   
95                0.0      0.0                 0.0           0.0   
96                0.0      0.0                 0.0           0.0   
97                0.0      0.0                 0.0           0.0   
98                0.0      0.0                 0.0           0.0   
99                0.0      0.0                 0.0           0.0   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0              0.0              0.0               0.0                 0.02   
1              0.0              0.0               0.0                 0.00   
2              0.0              0.0               0.0                 0.00   
3              0.0              0.0               0.0                 0.00   
4              0.0              0.0               0.0                 0.00   
..             ...              ...               ...                  ...   
95             0.0              0.0               0.0                 0.00   
96             0.0              0.0               0.0                 0.00   
97             0.0              0.0               0.0                 0.00   
98             0.0              0.0               0.0                 0.00   
99             0.0              0.0               0.0                 0.00   

    ...  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0   ...                           0.02               0.0          0.0   
1   ...                           0.00               0.0          0.0   
2   ...                           0.00               0.0          0.0   
3   ...                           0.00               0.0          0.0   
4   ...                           0.00               0.0          0.0   
..  ...                            ...               ...          ...   
95  ...                           0.00               0.0          0.0   
96  ...                           0.00               0.0          0.0   
97  ...                           0.00               0.0          0.0   
98  ...                           0.00               0.0          0.0   
99  ...                           0.00               0.0          0.0   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                     0.0              0.0      0.01        0.0          0.0   
1                     0.0              0.0      0.00        0.0          0.0   
2                     0.0              0.0      0.00        0.0          0.0   
3                     0.0              0.0      0.00        0.0          0.0   
4                     0.0              0.0      0.00        0.0          0.0   
..               

In [43]:

num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.07
1  Thai Restaurant  0.04
2             Café  0.04
3              Bar  0.04
4       Restaurant  0.04


----Agincourt----
                       venue  freq
0                     Lounge  0.25
1             Breakfast Spot  0.25
2  Latin American Restaurant  0.25
3               Skating Rink  0.25
4   Mediterranean Restaurant  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                      venue  freq
0                      Park   0.5
1                Playground   0.5
2  Mediterranean Restaurant   0.0
3         Mobile Phone Shop   0.0
4        Miscellaneous Shop   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                  venue  freq
0         Grocery Store  0.18
1          Liquor Store  0.09
2        Discount Store  0.09
3  Fast Food Restaurant  0.09
4              Pharmacy  0.09


----Alderwood,

In [44]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [45]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighborhood  \
0                           Adelaide, King, Richmond   
1                                          Agincourt   
2  Agincourt North, L'Amoreaux East, Milliken, St...   
3  Albion Gardens, Beaumond Heights, Humbergate, ...   
4                             Alderwood, Long Branch   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Coffee Shop       Thai Restaurant            Restaurant   
1  Latin American Restaurant        Breakfast Spot          Skating Rink   
2                       Park            Playground      Doner Restaurant   
3              Grocery Store          Liquor Store   Fried Chicken Joint   
4                Pizza Place           Coffee Shop                   Gym   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                  Café                   Bar      Sushi Restaurant   
1                Lounge             Drugstore        Discount Store   
2          Dessert Shop    Dim Sum Restaurant                 Diner   
3  Fast Food Restaurant              Pharmacy           Pizza Place   
4          Skating Rink                   Pub              Pharmacy   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0            Steakhouse        Breakfast Spot    Seafood Restaurant   
1   Distribution Center               Dog Run      Doner Restaurant   
2        Discount Store   Distribution Center               Dog Run   
3        Discount Store            Beer Store   Japanese Restaurant   
4        Sandwich Place                  Pool        Discount Store   

  10th Most Common Venue  
0                  Hotel  
1             Donut Shop  
2             Donut Shop  
3         Sandwich Place  
4           Dessert Shop

In [46]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_ 
# to change use .astype()

array([1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 3, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 4, 0, 1, 3, 0, 3, 4, 1, 1,
       1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 0, 1, 3, 1, 1, 1,
       1, 2, 1, 1, 1, 1, 3, 0, 3, 1, 0, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1,
       1, 3, 1, 0, 1, 3, 1, 0, 0, 1, 1, 3], dtype=int32)

In [47]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster_Labels', kmeans.labels_)

toronto_merged = Toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postalcode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             0.0  Fast Food Restaurant          Dessert Shop   
1 -79.160497             1.0                   Bar         Moving Target   
2 -79.188711             1.0                   Spa     Electronics Store   
3 -79.216917             0.0           Coffee Shop     Korean Restaurant   
4 -79.239476             1.0                Bakery           Gas Station   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Event Space  Ethiopian Restaurant   Empanada Restaurant   
1           Yoga Studio            Donut Shop        Discount Store   
2        Breakfast Spot   Rental Car Location        Medical Center   
3           Yoga Studio             Drugstore        Discount Store   
4      Hakka Restaurant   Fried Chicken Joint  Caribbean Restaurant   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
1   Distribution Center                      Dog Run      Doner Restaurant   
2    Mexican Restaurant                 Intersection           Yoga Studio   
3   Distribution Center                      Dog Run      Doner Restaurant   
4       Thai Restaurant           Athletics & Sports                  Bank   

  9th Most Common Venue       10th Most Common Venue  
0             Drugstore                   Donut Shop  
1             Drugstore           Dim Sum Restaurant  
2   Distribution Center                        Diner  
3            Donut Shop  Eastern European Restaurant  
4    Dim Sum Restaurant                        Diner

In [48]:
toronto_merged=toronto_merged.dropna()

In [49]:
toronto_merged['Cluster_Labels'] = toronto_merged.Cluster_Labels.astype(int)

In [50]:
# create map
map_clusters = folium.Map(location=[latitude_toronto, longitude_toronto], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

**Examine Clusters**

In [51]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels  1st Most Common Venue  \
0        Scarborough               0   Fast Food Restaurant   
3        Scarborough               0            Coffee Shop   
6        Scarborough               0       Department Store   
15       Scarborough               0   Fast Food Restaurant   
24        North York               0           Home Service   
33        North York               0          Grocery Store   
34        North York               0  Portuguese Restaurant   
80              York               0    Fried Chicken Joint   
85  Downtown Toronto               0            Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Dessert Shop           Event Space  Ethiopian Restaurant   
3      Korean Restaurant           Yoga Studio             Drugstore   
6            Coffee Shop        Discount Store           Bus Station   
15           Coffee Shop    Chinese Restaurant         Grocery Store   
24         Grocery Store               Butcher              Pharmacy   
33          Liquor Store        Discount Store           Coffee Shop   
34           Coffee Shop          Intersection          Hockey Arena   
80            Restaurant           Coffee Shop  Fast Food Restaurant   
85          Burger Joint                  Park                   Gym   

   5th Most Common Venue        6th Most Common Venue  \
0    Empanada Restaurant            Electronics Store   
3         Discount Store          Distribution Center   
6             Hobby Shop  Eastern European Restaurant   
15           Pizza Place               Breakfast Spot   
24           Pizza Place                  Coffee Shop   
33            Donut Shop                        Diner   
34           Yoga Studio                        Diner   
80        Sandwich Place          Distribution Center   
85           Yoga Studio                         Café   

          7th Most Common Venue  8th Most Common Venue 9th Most Common Venue  \
0   Eastern European Restaurant    Dumpling Restaurant             Drugstore   
3                       Dog Run       Doner Restaurant            Donut Shop   
6             Electronics Store    Dumpling Restaurant             Drugstore   
15               Sandwich Place            Supermarket     Electronics Store   
24               Discount Store            Yoga Studio    Dim Sum Restaurant   
33          Distribution Center                Dog Run      Doner Restaurant   
34               Discount Store    Distribution Center               Dog Run   
80                 Dessert Shop     Dim Sum Restaurant                 Diner   
85                          Bar  Portuguese Restaurant  Fast Food Restaurant   

         10th Most Common Venue  
0                    Donut Shop  
3   Eastern European Restaurant  
6                  Dessert Shop  
15                     Pharmacy  
24                        Diner  
33          Dumpling Restaurant  
34             Doner Restaurant  
80               Discount Store  
85                     Beer Bar

In [52]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
1    Scarborough               1                   Bar         Moving Target   
2    Scarborough               1                   Spa     Electronics Store   
4    Scarborough               1                Bakery           Gas Station   
5    Scarborough               1            Playground           Yoga Studio   
7    Scarborough               1                Bakery              Bus Line   
..           ...             ...                   ...                   ...   
96    North York               1   Empanada Restaurant           Pizza Place   
99     Etobicoke               1           Pizza Place        Sandwich Place   
100    Etobicoke               1     Mobile Phone Shop           Pizza Place   
101    Etobicoke               1         Grocery Store          Liquor Store   
102    Etobicoke               1   Rental Car Location             Drugstore   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1             Yoga Studio            Donut Shop        Discount Store   
2          Breakfast Spot   Rental Car Location        Medical Center   
4        Hakka Restaurant   Fried Chicken Joint  Caribbean Restaurant   
5              Donut Shop    Dim Sum Restaurant                 Diner   
7           Metro Station          Intersection           Bus Station   
..                    ...                   ...                   ...   
96            Yoga Studio      Doner Restaurant    Dim Sum Restaurant   
99            Coffee Shop          Intersection        Discount Store   
100        Sandwich Place              Bus Line           Yoga Studio   
101   Fried Chicken Joint  Fast Food Restaurant              Pharmacy   
102           Yoga Studio      Doner Restaurant    Dim Sum Restaurant   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1          Distribution Center               Dog Run      Doner Restaurant   
2           Mexican Restaurant          Intersection           Yoga Studio   
4              Thai Restaurant    Athletics & Sports                  Bank   
5               Discount Store   Distribution Center               Dog Run   
7               Ice Cream Shop          Soccer Field                  Park   
..                         ...                   ...                   ...   
96                       Diner        Discount Store   Distribution Center   
99   Middle Eastern Restaurant    Chinese Restaurant           Yoga Studio   
100        Distribution Center    Dim Sum Restaurant                 Diner   
101                Pizza Place        Discount Store            Beer Store   
102                      Diner        Discount Store   Distribution Center   

    9th Most Common Venue 10th Most Common Venue  
1               Drugstore     Dim Sum Restaurant  
2     Distribution Center                  Diner  
4      Dim Sum Restaurant                  Diner  
5        Doner Restaurant              Drugstore  
7            Costume Shop        Coworking Space  
..                    ...                    ...  
96                Dog Run              Drugstore  
99    Distribution Center     Dim Sum Restaurant  
100        Discount Store                Dog Run  
101   Japanese Restaurant         Sandwich Place  
102               Dog Run             Donut Shop  

[74 rows x 12 columns]

In [53]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
21  North York               2                   Gym      Department Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
21  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
21  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
21            Donut Shop       Doner Restaurant

In [54]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
14       Scarborough               3                  Park   
20        North York               3                  Park   
23        North York               3                  Park   
25        North York               3                  Park   
30        North York               3                  Park   
40         East York               3                  Park   
44   Central Toronto               3                  Park   
50  Downtown Toronto               3                  Park   
64   Central Toronto               3                  Park   
74              York               3                  Park   
79        North York               3                  Park   
90         Etobicoke               3                  Park   
98              York               3                  Park   

   2nd Most Common Venue       3rd Most Common Venue    4th Most Common Venue  \
14            Playground            Doner Restaurant             Dessert Shop   
20     Martial Arts Dojo                   Cafeteria             Dessert Shop   
23     Convenience Store                        Bank              Flower Shop   
25     Food & Drink Shop            Doner Restaurant       Dim Sum Restaurant   
30           Snack Place                     Airport         Doner Restaurant   
40           Coffee Shop           Convenience Store               Donut Shop   
44           Swim School                    Bus Line                  Dog Run   
50            Playground                       Trail                  Dog Run   
64         Jewelry Store            Sushi Restaurant                    Trail   
74         Women's Store                      Market     Ethiopian Restaurant   
79                Bakery  Construction & Landscaping                Drugstore   
90                  Pool                       River  Comfort Food Restaurant   
98     Convenience Store                  Donut Shop       Dim Sum Restaurant   

   5th Most Common Venue        6th Most Common Venue  \
14    Dim Sum Restaurant                        Diner   
20   Empanada Restaurant            Electronics Store   
23     Electronics Store  Eastern European Restaurant   
25                 Diner               Discount Store   
30    Dim Sum Restaurant                        Diner   
40                 Diner               Discount Store   
44    Dim Sum Restaurant                        Diner   
50          Dessert Shop           Dim Sum Restaurant   
64               Dog Run           Dim Sum Restaurant   
74   Empanada Restaurant            Electronics Store   
79        Discount Store          Distribution Center   
90      Department Store          Empanada Restaurant   
98                 Diner               Discount Store   

          7th Most Common Venue        8th Most Common Venue  \
14               Discount Store          Distribution Center   
20  Eastern European Restaurant          Dumpling Restaurant   
23          Empanada Restaurant          Dumpling Restaurant   
25          Distribution Center                      Dog Run   
30               Discount Store          Distribution Center   
40          Distribution Center                      Dog Run   
44               Discount Store          Distribution Center   
50                        Diner               Discount Store   
64                        Diner               Discount Store   
74  Eastern European Restaurant          Dumpling Restaurant   
79                      Dog Run             Doner Restaurant   
90            Electronics Store  Eastern European Restaurant   
98          Distribution Center                      Dog Run   

   9th Most Common Venue 10th Most Common Venue  
14               Dog Run             Donut Shop  
20             Drugstore             Donut Shop  
23    Dim Sum Restaurant              Drugstore  
25            Donut Shop     Falafel Restaurant  
30               Dog Run              Drugstore  
40      Doner

In [55]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
32  North York               4      Business Service            Food Truck   
91   Etobicoke               4      Business Service        Baseball Field   
97  North York               4        Baseball Field           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
32        Baseball Field           Yoga Studio      Doner Restaurant   
91           Yoga Studio            Donut Shop        Discount Store   
97            Donut Shop                 Diner        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32        Discount Store   Distribution Center               Dog Run   
91   Distribution Center               Dog Run      Doner Restaurant   
97   Distribution Center               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
32            Donut Shop     Falafel Restaurant  
91             Drugstore         Farmers Market  
97             Drugstore         Farmers Market